### PR TITLE
DDP Seeding Across Processes

### PR DESCRIPTION
**Pull request by ajaysaini725**
This change addresses issue #12 and introduces the following:
- Even when the user provides a seed, each process will have a different seed which is `user_provided_seed + global_rank`. The reason for this is that for algorithms like mixup which modify training batches, we want the modifications to be different on batches on different processes.
- When checkpointing, we store and load seeds using all processes, not just the seed from rank 0.

Test Plan:
- The existing distributed checkpointing tests cover the new code paths introduced here.

### CODE REVIEW

File: composer/trainer/checkpoint.py [hunk 1]
```diff
@@ -36,7 +36,11 @@ def load_checkpoint(self, state: State):
         self.checkpoint_rng_state = self._get_checkpoint_rng_state(state, self.state_dict["rng"])
 
         if "seed" in self.state_dict:
-            seed_all(self.state_dict["seed"])
+            assert ddp.get_world_size() == len(self.state_dict["seed"]), \
```
Comment (ravi-mosaicml) [inline composer/trainer/checkpoint.py:39]: I would make this into a warning. It's possible that we're restoring on different hardware

File: composer/trainer/trainer.py [hunk 2]
```diff
@@ -179,10 +179,12 @@ def __init__(
         self.device = device
 
         if not seed:
-            # Set a deterministic seed in the hparams
-            # This seed will be dumped in the hparams that are saved with checkpoints
             seed = get_random_seed()
             log.info(f"Seed was None. Setting seed to random value: {seed}")
+
+        # Assure that each process has a different seed, necessary if a seed is passed to init
+        seed += ddp.get_global_rank()
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:186]: When resuming from a checkpoint, would the global rank effectively be added twice (once upon the first training run, and again when resuming?)

Comment (ajaysaini725) [reply composer/trainer/trainer.py:186]: Nope because when restoring from checkpoint the seed is loaded from the save file in the `load_checkpoint` function and then `seed_all` is called with that seed so this codepath is nullified.

File: composer/trainer/checkpoint.py (hunk 1 above)
Comment (ajaysaini725) [reply composer/trainer/checkpoint.py:39]: The issue here is that you won't be able to restore the seeds correctly if the `world_size` doesn't match. If the new `world_size` is smaller but if it is larger then you won't be able to set seeds for the excess processes because there won't be any stored. 

Also, in general if the new `world_size` isn't the same then there can't be any deterministic guarantees when resuming from checkpoint in the case where a seed is set.

Once thing I can do is if the `world_size` doesn't match, then just log a warning and don't restore any seeds - does that sound good @ravi-mosaicml ?

File: composer/trainer/trainer.py [hunk 3]
```diff
@@ -308,7 +310,10 @@ def __init__(
             if self.deepspeed_enabled:
                 raise NotImplementedError("Checkpointing is not yet supported with DeepSpeed.")
             self.checkpoint_loader = CheckpointLoader(checkpoint_filepath=checkpoint_filepath)
-            self.checkpoint_loader.load_checkpoint(state=self.state)
+            restored_seed = self.checkpoint_loader.load_checkpoint(state=self.state)
+            if restored_seed is not None:
```
Comment (ajaysaini725) [inline composer/trainer/trainer.py:314]: This was necessary to handle the case:
- Train a model and save a checkpoint which includes the seed
- Resume from checkpoint and use the saved seed to seed all rng objects
- Save a checkpoint again (in this case `self.seed` is stored in the checkpoint so it must be the correct (i.e. originally checkpointed) seed
- Resume from checkpoint

File: composer/trainer/checkpoint.py (hunk 1 above)
Comment (ravi-mosaicml) [reply composer/trainer/checkpoint.py]: Yup, this is fine. I believe the existing behavior is to log a warning and not restore seeds.

File: composer/trainer/checkpoint.py [hunk 4]
```diff
@@ -36,7 +36,15 @@ def load_checkpoint(self, state: State):
...
-            seed_all(self.state_dict["seed"])
+            world_size = ddp.get_world_size()
+            checkpointed_world_size = len(self.state_dict["seed"])
+            if world_size != checkpointed_world_size:
+                warnings.warn(f"Current world size {world_size} does not match the checkpointed world size "
+                              f"{checkpointed_world_size}. The seed will not be restored.")
+                return
+            seed_to_restore = self.state_dict["seed"][ddp.get_global_rank()]
+            seed_all(seed_to_restore)
+            return seed_to_restore
```
Comment (ravi-mosaicml) [inline composer/trainer/checkpoint.py:47]: Can you update the docstring for the function?

File: composer/trainer/trainer.py (hunk 3 above)
Comment (ravi-mosaicml) [reply composer/trainer/trainer.py:314]: This is really helpful...I'd add this as a comment in the code

**Review by ravi-mosaicml — APPROVED** [review]
LGTM, though would really like to see some tests just verifiying this functionality...I don't think it should be too hard to extend the checkpointer tests with another parameterize.
